subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c

```diff
@@ -0,0 +1,1322 @@
+		return;
+	}
+
+	if (p->message.topic.qos == MQTT_QOS_1_AT_LEAST_ONCE) {
+		const struct mqtt_puback_param ack = {
+			.message_id = p->message_id
+		};
+
+		err = mqtt_publish_qos1_ack(client, &ack);
+		if (err) {
+			LOG_WRN("Failed to send MQTT ACK, error: %d",
+				err);
+		}
+	}
+
+#if IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS)
+	if (dps_reg_in_progress()) {
+		if (dps_process_message(&evt)) {
+			/* The message has been processed */
+			return;
+		}
+	}
+#endif
+
+	/* Check if it's a direct method invocation */
+	if (is_direct_method(topic, topic_len)) {
+		if (direct_method_process((char *)topic, topic_len,
+						payload_buf, payload_len)) {
+			LOG_DBG("Direct method processed");
+			return;
+		}
+
+		LOG_WRN("Unhandled direct method invocation");
+		return;
+	}
+
+	/* Check if message is a desired device twin change */
+	if (is_device_twin_update(topic, topic_len)) {
+#if IS_ENABLED(CONFIG_AZURE_FOTA)
+		err = azure_fota_msg_process(payload_buf, payload_len);
+		if (err < 0) {
+			LOG_ERR("Failed to process FOTA message");
+			return;
+		} else if (err == 1) {
+			LOG_DBG("Device twin update handled (FOTA)");
+			return;
+		}
+#endif /* IS_ENABLED(CONFIG_AZURE_FOTA) */
+		evt.type = AZURE_IOT_HUB_EVT_TWIN_DESIRED_RECEIVED;
+
+		azure_iot_hub_notify_event(&evt);
+	} else if (is_device_twin_result(topic, topic_len)) {
+		LOG_DBG("Device twin data received");
+
+		device_twin_result_process(topic, topic_len, payload_buf,
+					   payload_len);
+	} else {
+		azure_iot_hub_notify_event(&evt);
+	}
+}
+
+static void mqtt_evt_handler(struct mqtt_client *const client,
+			     const struct mqtt_evt *mqtt_evt)
+{
+	int err;
+	struct azure_iot_hub_evt evt = { 0 };
+
+	switch (mqtt_evt->type) {
+	case MQTT_EVT_CONNACK:
+		if (mqtt_evt->param.connack.return_code !=
+		    MQTT_CONNECTION_ACCEPTED) {
+			LOG_ERR("Connection was rejected with return code %d",
+				mqtt_evt->param.connack.return_code);
+			LOG_WRN("Is the device certificate valid?");
+				return;
+		}
+
+		connection_state = STATE_CONNECTED;
+
+		LOG_DBG("MQTT client connected");
+
+#if IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS)
+		if (dps_reg_in_progress()) {
+			err = dps_subscribe();
+			evt.type = AZURE_IOT_HUB_EVT_DPS_REGISTERING;
+		} else
+#endif
+		{
+			err = topic_subscribe();
+			evt.type = AZURE_IOT_HUB_EVT_CONNECTED;
+		}
+
+		if (err) {
+			LOG_ERR("Failed to subscribe to topics, err: %d", err);
+		}
+
+		azure_iot_hub_notify_event(&evt);
+		break;
+	case MQTT_EVT_DISCONNECT:
+		LOG_DBG("MQTT_EVT_DISCONNECT: result = %d", mqtt_evt->result);
+
+		connection_state = STATE_INIT;
+		evt.type = AZURE_IOT_HUB_EVT_DISCONNECTED;
+
+#if IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS)
+		if (dps_reg_in_progress()) {
+			if (dps_process_message(&evt)) {
+				/* The message has been processed */
+				break;
+			}
+		}
+#endif
+		if (atomic_get(&dps_disconnecting) == 1) {
+			break;
+		}
+
+		azure_iot_hub_notify_event(&evt);
+		break;
+	case MQTT_EVT_PUBLISH: {
```

Comment:
nit: I don't see a reason to add braces here (would be consistent with the rest of the code).

Reply:
Fixed, not needed after stuff was moved into `on_publish()`